src/crimson/net/SocketMessenger.cc

```diff
@@ -147,10 +153,16 @@ seastar::future<> SocketMessenger::shutdown()
   if (listener) {
     listener->abort_accept();
   }
+  // close all connections
   return seastar::parallel_for_each(connections.begin(), connections.end(),
     [this] (auto conn) {
       return conn.second->close();
-    }).finally([this] { connections.clear(); });
+    }).finally([this] {
+      connections.clear();
+      // closing connections will unblock any dispatchers that were waiting to
+      // send(). wait for any pending calls to ms_dispatch() to finish
+      return std::move(pending_dispatch);
```

Comment:
We also need to wait for other dispatching tasks before shutdown, right?
Maybe we could implement a dedicated class to collect dispatching futures, trim them periodically, and provide a safety future to allow a graceful shutdown. (TODO?)

Reply:
yeah, futures from ms_handle_accept/connect/etc should be managed too